### PR TITLE
network: rework address pool

### LINKED ISSUE #9955: Automatic address allocation does not work for IPv6 [::]
**systemd version the issue has been seen with**
> 231 - 238

**Used distribution**
Debian, Gentoo and CoreOS

**Expected behaviour you didn't see**
 > I get an fc00::/7 auto generated ULA IP address assigned

**Unexpected behaviour you saw**
 > Various errors and crashes (see below), generally not working.

**Steps to reproduce the problem**
As I tried this various times on my live gentoo server, I switched to using a VM as I couldn't get it to work. I tried it on my Debian system (buster, desktop), but blamed potentially network-manager interfering (which turns out was not the case).
I booted up an 'alpha' version of coreos in a VM to ensure I could get the latest possible systemd version, 238.

So I made various versions of /etc/systemd/network/30-eth0.network with the following results, here the first version:
`
[Match]
MACAddress=52:54:00:*

[Network]
LinkLocalAddressing=yes
Description="IPv6 test"
DHCP=yes
Address=[::]
Address=0.0.0.0
`
(I added the ipv4 address to see if it was working at all, for the below variations, I left it out)

The above config results in the following:
```
Aug 27 19:38:49 localhost systemd-networkd[915]: /etc/systemd/network/30-eth0.network:8: Address is invalid, ignoring assignment: [::]
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Link 2 added
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: udev initialized link
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Saved original MTU: 1500
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: fe80::5054:ff:fe76:4f86/64 (valid forever)
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Gained IPv6LL
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: 10.2.10.138/16 (valid for 6min 5s)
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: 10.0.0.1/8 (valid forever)
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: 169.254.6.149/16 (valid forever)
```

As can be seen, [::] is being ignored, despite what's being written in the [manual](https://github.com/systemd/systemd/blob/master/man/systemd.network.xml#L484)

When looking at the [source](https://github.com/systemd/systemd/blob/master/src/network/networkd-address.c#L746) which causes the error, it appears that this is triggered from the configuration file parser. The documented [::] [parser](https://github.com/systemd/systemd/blob/master/src/network/networkd-address.c#L479) or at least notion off, only seems to be called from other functions which don't seem to related to the address parsing from file:
```
src/network/networkd-dhcp4.c:     r = address_configure(addr, link, dhcp4_address_handler, true);
src/network/networkd-dhcp6.c:     r = address_configure(addr, link, dhcp6_address_handler, true);
src/network/networkd-ipv4ll.c:       r = address_configure(ll_addr, link, ipv4ll_address_handler, false);
src/network/networkd-link.c:          r = address_configure(ad, link, address_handler, false);
src/network/networkd-ndisc.c:       r = address_configure(address, link, ndisc_netlink_handler, true);
```
so I'm inclined to think that's less relevant.

Going back to the configuration file parser, the failing function, in_addr_from_string_auto, is really just a wrapper around inet_pton, which is does state that :: gets [silently ignored](https://github.com/git/git/blob/master/compat/inet_pton.c#L97) (note that this is the git compat inet_pton, couldn't find a quick reference of the real one).

What I think happens is, that we tell the user to enter [::] which gets passed to inet_ptons which ignores the [ and thus returns invalid.

Switching to Address=::, which seems logical as inet_pton would accept that, yields not [unexpected](https://github.com/systemd/systemd/blob/master/src/network/networkd-manager.c#L1685) an error:
`
Assertion 'prefixlen > 0' failed at ../systemd-238/src/network/networkd-manager.c:1698, function manager_address_pool_acquire(). Aborting.
`

Of course, trying the next permutation, Address=::/7 'works' by satisfying both the prefixlen requirement and the inet_pton requirement, but not satisfactory as it sets the IPv6 address to fc00::1/7.

So either the documentation is wrong, in that ::/7 should be used, but then in turn, the generated IP address is wrong, as the fc00/7 subnet should generate a [ULA](https://en.wikipedia.org/wiki/Unique_local_address) IP, which more or less forces the IP into the fd00/8 subnet due to the required bit having to be set and also be randomly generated as per [RFC4193](https://tools.ietf.org/html/rfc4193) and thus ::/7 currently generates non-compliant address.

Or the implementation is wrong, in that [::] is passed straight to inet_pton, which is wrong as it needs the [ (and ]) stripped so that inet_pton happily approves the general IP and in turn, systemd can generate the required ULA IP using the random algorithm as described by [RFC4193](https://tools.ietf.org/html/rfc4193). But I haven't traced that bit out if systemd even those that in any case.

In either case, the manual does state, that [::] should [generate](https://github.com/systemd/systemd/blob/master/man/systemd.network.xml#L484) address and I cannot imagine it not being compliant.

P.S. I apologize beforehand for repeating and overstating things, it's quite late and I'm really tired ;)

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #9955.

### CODE REVIEW

**Comment by yuwata** [comment]
Updated. PTAL.

cc @ssahani.

File: src/network/networkd-manager.c [hunk 1]
```diff
@@ -38,7 +38,7 @@ static int setup_default_address_pool(Manager *m) {
 
         /* Add in the well-known private address ranges. */
 
-        r = address_pool_new_from_string(m, &p, AF_INET6, "fc00::", 7);
+        r = address_pool_new_from_string(m, &p, AF_INET6, "fd00::", 8);
```
Comment (poettering) [inline src/network/networkd-manager.c:42]: so what does this mean precisely? that only half the previous range is now among the default? is that correct?

File: src/basic/in-addr-util.c [hunk 2]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
...
+
+                if (prefixlen_fixed_part > 32)
+                        prefixlen_fixed_part = 32;
+                if (prefixlen > 32)
+                        prefixlen = 32;
+                if (prefixlen_fixed_part >= prefixlen)
+                        return -EINVAL;
+
+                c = be32toh(u->in.s_addr);
+                c &= ((1UL << (prefixlen_fixed_part + 1)) - 1) << (32 - prefixlen_fixed_part);
```
Comment (poettering) [inline src/basic/in-addr-util.c]: 1UL → UINT32_C(1), no?

(After all UL on x86-64 is 64 bit which isn't what you want here)

File: src/basic/in-addr-util.c [hunk 3]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
...
+
+                random_bytes(&n, sizeof(n));
+                n ^= ((1UL << (prefixlen - prefixlen_fixed_part + 1)) - 1) << (32 - prefixlen);
```
Comment (poettering) [inline src/basic/in-addr-util.c]: dito

File: src/basic/in-addr-util.c [hunk 4]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
...
+
+                for (i = 0; i < 16; i++) {
+                        uint8_t mask_fixed_part = 0, mask = 0;
+
+                        if (i < (prefixlen_fixed_part + 7) / 8) {
+                                if (i < prefixlen_fixed_part / 8)
+                                        mask_fixed_part = 0xff;
+                                else {
+                                        j = prefixlen_fixed_part % 8;
+                                        mask_fixed_part = ((1UL << (j + 1)) - 1) << (8 - j);
```
Comment (poettering) [inline src/basic/in-addr-util.c]: 1UL → UINT8_C(1)?

File: src/basic/in-addr-util.c [hunk 5]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
...
+                                }
+                        }
+
+                        if (i < (prefixlen + 7) / 8) {
+                                if (i < prefixlen / 8)
+                                        mask = 0xff ^ mask_fixed_part;
+                                else {
+                                        j = prefixlen % 8;
+                                        mask = (((1UL << (j + 1)) - 1) << (8 - j)) ^ mask_fixed_part;
```
Comment (poettering) [inline src/basic/in-addr-util.c]: dito

File: src/basic/in-addr-util.c [hunk 6]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
...
+                                        mask_fixed_part = 0xff;
+                                else {
+                                        j = prefixlen_fixed_part % 8;
+                                        mask_fixed_part = ((1UL << (j + 1)) - 1) << (8 - j);
+                                }
+                        }
+
+                        if (i < (prefixlen + 7) / 8) {
+                                if (i < prefixlen / 8)
+                                        mask = 0xff ^ mask_fixed_part;
```
Comment (poettering) [inline src/basic/in-addr-util.c]: 0xff → 0xffu?

File: src/basic/in-addr-util.c [hunk 7]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
...
+                        return -EINVAL;
+
+                random_bytes(&n, sizeof(n));
+
+                for (i = 0; i < 16; i++) {
+                        uint8_t mask_fixed_part = 0, mask = 0;
+
+                        if (i < (prefixlen_fixed_part + 7) / 8) {
+                                if (i < prefixlen_fixed_part / 8)
+                                        mask_fixed_part = 0xff;
```
Comment (poettering) [inline src/basic/in-addr-util.c]: 0xffu?

File: src/basic/in-addr-util.c [hunk 8]
```diff
@@ -215,6 +216,83 @@ int in_addr_prefix_next(int family, union in_addr_union *u, unsigned prefixlen)
         return -EAFNOSUPPORT;
 }
 
+int in_addr_random_prefix(
```
Comment (poettering) [inline src/basic/in-addr-util.c:219]: maybe sould have a test case of its own?

File: src/network/networkd-address-pool.c [hunk 9]
```diff
@@ -131,7 +134,19 @@ int address_pool_acquire(AddressPool *p, unsigned prefixlen, union in_addr_union
...
+        for (i = 0; p->family == AF_INET6 ? i < RANDOM_PREFIX_TRIAL_MAX : true; i++) {
+                if (p->family == AF_INET6) {
+                        _cleanup_free_ char *buf = NULL;
+
+                        r = in_addr_random_prefix(p->family, &u, p->prefixlen, prefixlen);
+                        if (r <= 0)
+                                return r;
+
+                        (void) in_addr_to_string(p->family, &u, &buf);
+                        log_debug("Address pool: Address %s/%u is generated.", strnull(buf), prefixlen);
```
Comment (poettering) [inline src/network/networkd-address-pool.c]: enclose these two lines and the _cleanup_free line above in `if (DEBUG_LOGGING)`?

File: src/network/networkd-address-pool.c [hunk 10]
```diff
@@ -131,7 +134,19 @@ int address_pool_acquire(AddressPool *p, unsigned prefixlen, union in_addr_union
                 return 0;
 
         u = p->in_addr;
-        for (;;) {
+
+        for (i = 0; p->family == AF_INET6 ? i < RANDOM_PREFIX_TRIAL_MAX : true; i++) {
+                if (p->family == AF_INET6) {
```
Comment (poettering) [inline src/network/networkd-address-pool.c]: hmm, why only do this for ipv6 btw? why not for ipv4 too?

File: src/network/networkd-manager.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-manager.c:42]: `fc00::/8` may be used in the future. From [rfc4193](https://tools.ietf.org/html/rfc4193#section-3.1), about the 8th bit,
> Set to 1 if the prefix is locally assigned.
> Set to 0 may be defined in the future.

**Comment by yuwata** [comment]
Updated. I hope all comments are addressed. PTAL.
- now IPv4 addresses are also generated with random prefix.
- tests for `in_addr_random_prefix()` are added.

File: src/network/networkd-address.c [hunk 11]
```diff
@@ -495,8 +495,9 @@ static int address_acquire(Link *link, Address *original, Address **ret) {
         assert(ret);
 
         /* Something useful was configured? just use it */
-        if (in_addr_is_null(original->family, &original->in_addr) <= 0)
-                return 0;
+        r = in_addr_is_null(original->family, &original->in_addr);
+        if (r <= 0)
+                return r;
```
Comment (poettering) [inline src/network/networkd-address.c:500]: hmm, this function logs about all other errors. this should log about the r < 0 cases here too, no?

Comment (poettering) [reply src/network/networkd-address.c:500]: hmm, actually it's chaos right now. This should probably be cleared up, but probably doesn't have to be in this PR. But in the long run we should make sure that each function either logs about all errors or none, and this one is a bit messy there right now, as it logs about failures to acquire addresses from the pool, but not on -ENOMEM for example...

but anyway, given that this patch doesn't really make things worse, let's ignore this for this PR

Comment (yuwata) [reply src/network/networkd-address.c:500]: Theoretically yes. I hope that the family is already checked at somewhere before calling this function. But yes, I will add the logs.

**Pull request merged**